### PR TITLE
Implemented CupertinoButton new styles/sizes (fixes #92525)

### LINKED ISSUE #92525: [Cupertino] Add new iOS 15 button styles for `CupertinoButton`
Currently, there are only two styles aka button [configurations](URLhttps://developer.apple.com/documentation/uikit/uibutton/configuration)  for `CupertinoButton`
Plain [CupertinoButton](https://master-api.flutter.dev/flutter/cupertino/CupertinoButton/CupertinoButton.html) & [CupertinoButton.filled](https://master-api.flutter.dev/flutter/cupertino/CupertinoButton/CupertinoButton.filled.html)

Add following new button configurations  
[Gray](https://developer.apple.com/documentation/uikit/uibutton/configuration/3750787-gray)
[Tinted](https://developer.apple.com/documentation/uikit/uibutton/configuration/3750798-tinted)

### PR DESCRIPTION
**Pull request by kerberjg**
This PR fixes #92525 and introduces the following changes according to [latest iOS HIG](https://developer.apple.com/design/human-interface-guidelines/buttons#iOS-iPadOS):

- `CupertinoButton` now has a `size` property (type `enum CupertinoButtonSize`, values sm/md/lg, default `lg`) that allows the devs to apply new iOS 15+ button styles
- Previously `CupertinoButton` had a larger padding when no background color was specified. With the new HIG, that is no longer the case
- `CupertinoButton` now has a `.tinted` constructor that renders a translucent background (transparency % is brightness-dependent) and uses a different foreground color compared to `.filled`
- `CupertinoButton` now uses the `actionTextStyle` TextStyle from the given theme
- `CupertinoButton`'s child IconTheme's size will always be x1.2 the given TextStyle's size
- `CupertinoTextThemeData` now has a `actionSmallTextStyle` property to use with small buttons (including a default `_kDefaultActionSmallTextStyle` TextStyle)

Preview & example:

image

> **NOTE**: there is a discrepancy in dark mode button foreground color between the default CupertinoTheme and the HIG. A separate issue will be opened for this.

~EDIT: issue reported here https://github.com/flutter/flutter/issues/152846~
EDIT2: fixed by #153039 !

image

## Example
```dart
import 'package:flutter/cupertino.dart';

const Widget body = Row(
  mainAxisSize: MainAxisSize.min,
  mainAxisAlignment: MainAxisAlignment.center,
  children: <Widget>[
    Icon(
      CupertinoIcons.play_fill,
    ),
    Text("Play"),
  ],
);

void main() =>
  runApp(
    CupertinoApp(
      home: Container(
        child: Wrap(
        direction: Axis.horizontal,
        children: <Widget>[
          // header
          Text(''),
          Text('Plain'),
          Text('Grey'),
          Text('Tinted'),
          Text('Filled'),
          // small
          Text('Small'),
          CupertinoButton(
            child: body,
            onPressed: () {},
            size: CupertinoButtonSize.small,
          ),
          CupertinoButton.tinted(
            child: body,
            onPressed: () {},
            size: CupertinoButtonSize.small,
            color: CupertinoColors.systemGrey,
          ),
          CupertinoButton.tinted(
            child: body,
            onPressed: () {},
            size: CupertinoButtonSize.small,
          ),
          CupertinoButton.filled(
            child: body,
            onPressed: () {},
            size: CupertinoButtonSize.small,
          ),
          // medium
          Text('Medium'),
          CupertinoButton(
            child: body,
            onPressed: () {},
            size: CupertinoButtonSize.medium,
          ),
          CupertinoButton.tinted(
            child: body,
            onPressed: () {},
            size: CupertinoButtonSize.medium,
            color: CupertinoColors.systemGrey,
          ),
          CupertinoButton.tinted(
            child: body,
            onPressed: () {},
            size: CupertinoButtonSize.medium,
          ),
          CupertinoButton.filled(
            child: body,
            onPressed: () {},
            size: CupertinoButtonSize.medium,
          ),
          // large
          Text('Large'),
          CupertinoButton(
            child: body,
            onPressed: () {},
            size: CupertinoButtonSize.large,
          ),
          CupertinoButton.tinted(
            child: body,
            onPressed: () {},
            color: CupertinoColors.systemGrey,
            size: CupertinoButtonSize.large,
          ),
          CupertinoButton.tinted(
            child: body,
            onPressed: () {},
            size: CupertinoButtonSize.large,
          ),
          CupertinoButton.filled(
            child: body,
            onPressed: () {},
            size: CupertinoButtonSize.large,
          ),
        ].map((Widget w) => SizedBox(width: 110, height: 70, child: Center(child: w))).toList(),
      ),
      )
    ),
  );

```

*List which issues are fixed by this PR. You must list at least one issue. An issue is not required if the PR fixes something trivial like a typo.*

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

File: packages/flutter/lib/src/cupertino/button.dart
Comment (MitchellGoodwin) [inline packages/flutter/lib/src/cupertino/button.dart:29]: These maps that store raw style values should probably be made public. It makes it easier for developers making custom widgets to use, either by just referencing them in the docs or pulling them in directly in the code.

Comment (MitchellGoodwin) [reply packages/flutter/lib/src/cupertino/button.dart:29]: We tend to make these kind of values private, and it's usually good practice to do so for writing app code, but we've received feedback that in the framework code it's a tad annoying.

File: packages/flutter/lib/src/cupertino/button.dart
Comment (MitchellGoodwin) [inline packages/flutter/lib/src/cupertino/button.dart:223]: Dart doc doesn't pick up bracket references to private references well. No issue if we make `_kCupertinoButtonSizeBorderRadius` public.

**Review by MitchellGoodwin — COMMENTED** [review]
These look great! I did a first pass at the implementation without going too deep into the details, and made some comments.

I like all of the styling values be in maps for the different size styles at the top. It makes it easy to look through and update the values later if HIGs changes. Just throwing an idea out here, but I wonder if instead of splitting up maps by property type, like `_kCupertinoButtonSizeBorderRadius` and `_kCupertinoButtonMinSize`, we instead of have something like Material's `ButtonStyle`, and we have something like a `CupertinoButtonStyle` class that contains all of the properties for each configuration in one place, for the different sizes as well as filled, outlined and so on. It might be a little bit easier to search through or use in other widgets if needed. The button could then have a `_style` property that gets set based on the constructor and properties. For instance for a filled button of medium size, it does something like this
```
_style = CupertinoButtonStyle.filled.mergeWith(CupertinoButtonStyle.medium)
```
Then throughout the code it can just check with the one `_style` collection rather than having to do multiple checks for the button type. 
```
final Color? backgroundColor =  _style.backgroundColor
```

**Comment by kerberjg** [comment]
@MitchellGoodwin Thanks for the review! I agree it would be good to have a `CupertinoButtonStyle`, also because that's be closer to how iOS' `UIButton` works - I think it'd be worthwhile to open a separate issue/PR to handle that properly, and for now just introduce a minimal HIG-compliant version of the new properties in this one.

I addressed some of your feedback in the latest commits and left you some questions on others

File: packages/flutter/lib/src/cupertino/button.dart
Comment (MitchellGoodwin) [reply packages/flutter/lib/src/cupertino/button.dart:403]:
```suggestion
      size: ( textStyle.fontSize ?? 14 ) * 1.2,
```
14 would probably be an ok default. It *could* scale up or down depending on `sizeStyle` but that might be overengineering for the case where somebody deliberately removed theme values.

File: packages/flutter/lib/src/cupertino/button.dart
Comment (MitchellGoodwin) [inline packages/flutter/lib/src/cupertino/button.dart:30]: These would need some documentation, pointing out they are style defaults at different `[CupertinoButtonSize]` values. Also include that they come from the iOS 17 HIGs.

File: packages/flutter/test/cupertino/button_test.dart
Comment (MitchellGoodwin) [inline packages/flutter/test/cupertino/button_test.dart:31]: Is this going up because of the padding from `kCupertinoButtonPadding` for the large sized style?

**Review by MitchellGoodwin — COMMENTED** [review]
Looks like at least one test is failing for the button, and there's some analyzer issues. Take a look at the 'Linux Analyze' check for the analyzer. That check isn't too bad to look through, but let me know if you have issues finding other errors. The checks can be a slog to look through until you are used to it.

**Comment by MitchellGoodwin** [comment]
> for now just introduce a minimal HIG-compliant version of the new properties in this one.

Let's do that. On a second look, I think this implementation is more than sufficient.

File: packages/flutter/lib/src/cupertino/button.dart
Comment (victorsanni) [inline packages/flutter/lib/src/cupertino/button.dart:78]: These constants are used everywhere for focus in cupertino: switch, checkbox, radio, button, etc. If we are making them public, they should probably be defined in one place (possibly `constants.dart`) and made public there instead.

File: packages/flutter/lib/src/cupertino/button.dart
Comment (victorsanni) [inline packages/flutter/lib/src/cupertino/button.dart:23]: Nit that will inevitably come up at some point: all comments should end in a period.

File: packages/flutter/lib/src/cupertino/button.dart
Comment (victorsanni) [inline packages/flutter/lib/src/cupertino/button.dart:152]: Does this default disabled color also match native iOS in dark mode? If not, consider using CupertinoDynamicColor.resolve to make it resolve differently in light vs. dark modes.

File: packages/flutter/lib/src/cupertino/button.dart
Comment (MitchellGoodwin) [inline packages/flutter/lib/src/cupertino/button.dart:19]:
```suggestion
/// Based on the iOS (17) [Human Interface Guidelines](https://developer.apple.com/design/human-interface-guidelines/buttons#iOS-iPadOS).
```
nit: Prefer formatted links, plus punctuation.

File: packages/flutter/lib/src/cupertino/button.dart
Comment (MitchellGoodwin) [inline packages/flutter/lib/src/cupertino/button.dart:30]: Same here, and below for the similar constants.

File: packages/flutter/lib/src/cupertino/button.dart
Comment (MitchellGoodwin) [inline packages/flutter/lib/src/cupertino/button.dart:82]:
```suggestion
/// Opacity values for the background of a [CupertinoButton.tinted].
```
nit

File: packages/flutter/lib/src/cupertino/button.dart
Comment (victorsanni) [reply packages/flutter/lib/src/cupertino/button.dart:78]: I'm wondering in general where these public constants should live. Maybe the ones directly relevant to CupertinoButton should live here, and everything else move to cupertino/constants.dart?

File: packages/flutter/lib/src/cupertino/button.dart
Comment (MitchellGoodwin) [inline packages/flutter/lib/src/cupertino/button.dart:240]:
```suggestion
  /// The size of the button.
  ///
  /// Defaults to [CupertinoButtonSize.large].
```
nit

File: packages/flutter/lib/src/cupertino/button.dart
Comment (MitchellGoodwin) [inline packages/flutter/lib/src/cupertino/button.dart:435]: I wonder if it'd be a good idea to change get these values from a static function using a switch statement, rather than a map, so that we don't have to use the `!` symbol?.

**Review by MitchellGoodwin — COMMENTED** [review]
Thank you for the update! Left some comments.

File: packages/flutter/lib/src/cupertino/button.dart
Comment (victorsanni) [reply packages/flutter/lib/src/cupertino/button.dart:78]: Opened https://github.com/flutter/flutter/pull/153115 to kickstart this move.

File: packages/flutter/test/cupertino/button_test.dart
Comment (MitchellGoodwin) [inline packages/flutter/test/cupertino/button_test.dart:527]: Good idea checking for a regression here

File: packages/flutter/lib/src/cupertino/button.dart
Comment (MitchellGoodwin) [inline packages/flutter/lib/src/cupertino/button.dart:20]:
```suggestion
/// The size of a [CupertinoButton].
///
/// Based on the iOS (17) [Human Interface Guidelines](https://developer.apple.com/design/human-interface-guidelines/buttons#iOS-iPadOS).
```
Dart doc comments need a line between the first sentence and the rest. The first line is a short summary that shows up as a preview when the detailed view isn't needed. It makes it easier to read in the code when there's a gap. This needs to happen throughout this PR.

File: packages/flutter/test/cupertino/button_test.dart
Comment (MitchellGoodwin) [inline packages/flutter/test/cupertino/button_test.dart:528]:
```suggestion
    for (final CupertinoButtonSize size in CupertinoButtonSize.values) {
```
Analyzer is complaining about this

File: packages/flutter/test/cupertino/button_test.dart
Comment (MitchellGoodwin) [reply packages/flutter/test/cupertino/button_test.dart:527]: The analyzer looks like it's complaining about the `/'`

**Review by MitchellGoodwin — COMMENTED** [review]
Left some nits. This is getting close. Sorry if this is a slog, there's a lot of process around public style changes like this.

**Comment by kerberjg** [comment]
@MitchellGoodwin no worries! It's a good opportunity to learn, now I know what to watch out for next time :)

I rebased on main and pushed the latest changes based on your feedback.

Earlier the CI reported a mismatch with the goldens due to the default icon size change, shall I go ahead and just update them?

**Comment by MitchellGoodwin** [comment]
> Earlier the CI reported a mismatch with the goldens due to the default icon size change, shall I go ahead and just update them?

I think you need to be a member of flutter-hackers to be able to approve golden file differences, so one of us will approve them. Typically I like to wait until the PR is in a done state before I approve them. When I looked at them before the differences were all as expected.

File: packages/flutter/test/cupertino/button_test.dart
Comment (victorsanni) [inline packages/flutter/test/cupertino/button_test.dart:483]: Nit: extra newline here should be removed.

File: packages/flutter/test/cupertino/button_test.dart
Comment (victorsanni) [inline packages/flutter/test/cupertino/button_test.dart:712]: Nit: Comments should start with a capital letter and end with a period, as per the style guide.

File: packages/flutter/test/cupertino/button_test.dart
Comment (victorsanni) [inline packages/flutter/test/cupertino/button_test.dart:688]: Nit: Comments should start with a capital letter and end with a period, as per the style guide.

**Review by victorsanni — COMMENTED** [review]
A couple of nits, but this is close to landing. Thanks once again for the patience.

**Comment by kerberjg** [comment]
Nits addressed

<sup>*Deploy Fridays, let's go!*</sup>

File: packages/flutter/lib/src/cupertino/text_theme.dart
Comment (victorsanni) [inline packages/flutter/lib/src/cupertino/text_theme.dart:46]: Is there a specific figma file these values were derived from? If there is, I'd recommend to link it directly.

File: packages/flutter/test/cupertino/button_test.dart
Comment (victorsanni) [inline packages/flutter/test/cupertino/button_test.dart:157]: Should this be removed? We typically don't merge commented out code.

**Review by victorsanni — APPROVED** [review]

**Review by MitchellGoodwin — APPROVED** [review]
LGTM as well! Thank you very much!

**Comment by MitchellGoodwin** [comment]
Golden file differences were as expected from the different default styling.

**Comment by kerberjg** [comment]
Thank you too! It was good practice to get this one merged, on to the next one! 😊

**Pull request merged**
